feat: adds Java type generation.

pcornish opened the pull request:
fixes #72 

## What does this do

This PR adds a project that generates data model classes as Java POJOs, based on the JSON schema files.

The generated Java types are written to the `build/generated-sources/di-data-model` directory.

### Usage

Here is an example using the `IdentityCheckCredentialJWT` class.

> **Note**
> In this example, we make use of the Gang-of-Four builder pattern (the static `builder()` method, and its
> associated `with...` methods).
> 
> If preferred, the POJO's constructor, and its setters and getters can be used to construct
> instances instead. 

```java
var credentials = IdentityCheckCredentialJWT.builder()
    .withSub("urn:fdc:gov.uk:2022:954bc117-731b-41cd-86cf-dfb4e7940fce")
    .withAud("https://passport.core.stubs.account.gov.uk")
    .withNbf(1690816091)
    .withIss("https://review-p.build.account.gov.uk")
    .build();

var vc = IdentityCheckCredentialClass().builder()
    .withType(List.of(
        VerifiableCredentialType.VERIFIABLE_CREDENTIAL,
        VerifiableCredentialType.IDENTITY_CHECK_CREDENTIAL
    )).build();

credentials.setVc(vc);

var evidence = IdentityCheckClass.builder()
    .withValidityScore(0)
    .withStrengthScore(4)
    .withCi(List.of("D02"))
    .withTxn("5f57a8f2-62b0-4958-9332-06d9f453e5b9")
    .withType("IdentityCheck")
    .build();

vc.setEvidence(List.of(evidence));
```

## Example

See the unit test `uk.gov.di.model.ModelTest` (under `src/test/java`) for an example of using the
model, then serialising it to JSON.

## Build and run

### Prerequisites

- Generate the schema files from this project (see `scripts/generate_json_schemas.sh`)
- Ensure you have Java 17+ installed

## Steps

Generate the Java types and run the unit tests:

```shell
./gradlew build
```

The generated Java types are written to the `build/generated-sources/di-data-model` directory.

## Implementation notes

The class generation is done using the widely-used https://github.com/joelittlejohn/jsonschema2pojo project (specifically its Gradle plugin).

The model classes are annotated using Jackson (though other options are possible).

## Publishing

Like the TypeScript variant, the build pipeline produces and publishes an artifact consumable by dependency management tools.

The Maven artifact is published here: https://github.com/orgs/govuk-one-login/packages?repo_name=data-vocab

The Maven publishing configuration is [documented here](https://github.com/govuk-one-login/data-vocab/blob/feat/java-types/scripts/java-types/README.md#publish-maven-artifact).

pcornish committed:
feat: adds Java type generation.

pcornish committed:
ci: adds Java type generation to build pipeline.

pcornish committed:
test: switches to JSONAssert library for model test assertions.

pcornish committed:
feat(java-types): enables builder methods.

pcornish committed:
build(java-types): explicitly sets some generator options.

pcornish committed:
feat(java-types): only generate types from topmost schema file.

pcornish committed:
test(java-types): externalises JSON exemplar to file.

pcornish committed:
docs(java-types): adds readme.

pcornish committed:
feat(java-types): adds Maven publish build configuration.

pcornish committed:
ci(java-types): publishes Java types as Maven artifact.

fewtrees-pa reviewed (commented):
The principle is good with the following caveats:

The Builder pattern in https://github.com/joelittlejohn/jsonschema2pojo doesn’t generate builders but syntactic sugar to circumvent the bean method setter standard and is a bit of a smell.. I personally would not use it and use Lombok Builder annotations instead.

I'm not sure if adding the Json serialisation annotations is a good idea - it could want versions that do not comply/agree with the currently supported versions we use. I've had many headaches from cross-dependencies with json libs..

Will the tool handle multiple input schema files? There's only 1 output package name and each inout schema file also generates a `$defs` class - will it aggregate schema defs

What is really needed is a tool that converts json schema to a lombok annotated class (https://projectlombok.org/features/Builder) so all you need is:

```
@Builder
@Data
class MyObj {
  String firstName;
  String lastName;
}
```

..and then you get everything else for free..

pcornish committed:
refactor: co-locates type generators under common parent.

pcornish committed:
feat(java-types): splits POJO and annotated artifact generators.

pcornish committed:
build(java-types): moves common build config to parent module.

pcornish committed:
build(java-types): defined jackson dependency as a range.

pcornish committed:
feat(java-types): switches builder to Gang-of-Four type.

This builder type removes the 'with...' methods on the data model classes, making them more POJO-like. Instead, the builder uses a separate class, per the Gang-of-Four builder pattern.

pcornish committed:
build: release script updates Java types project version.

Joe-Edwards-GDS commented on code-generators/java-types/plain-pojo/src/test/resources/serialised_model.json:
```diff
@@ -0,0 +1,62 @@
+  "nbf" : 1690816091,
+  "sub" : "urn:fdc:gov.uk:2022:954bc117-731b-41cd-86cf-dfb4e7940fce",
+  "vc" : {
+    "context" : null,
```
Note that in the plain POJO this property name is incorrect - it should be `@context` (not `context`). It doesn't really matter since afaict we never actually populate this property - it's a W3 VC property https://www.w3.org/TR/vc-data-model/#contexts - but a potential gotcha

pcornish replied:
Good spot! This is resolved in the test now. The serialisation wasn't configured properly, but now the serialised JSON is the same between the annotated and plain modules.

pcornish committed:
test: switches to deserialisation and builder tests to be less brittle.

pcornish committed:
build: allows script root dir to be set by parent environment.

pcornish committed:
test: adds containerised build for isolated local generation.
